cpp/src/io/parquet/page_data.cu

```diff
@@ -47,395 +27,6 @@ namespace gpu {
-        run = (cur < end) ? get_vlq32(cur, end) : 0;
-        if (!(run & 1)) {
-          // Repeated value
-          int bytecnt = (dict_bits + 7) >> 3;
-          if (cur + bytecnt <= end) {
-            int32_t run_val = cur[0];
-            if (bytecnt > 1) {
-              run_val |= cur[1] << 8;
-              if (bytecnt > 2) {
-                run_val |= cur[2] << 16;
-                if (bytecnt > 3) { run_val |= cur[3] << 24; }
-              }
-            }
-            s->dict_val = run_val & ((1 << dict_bits) - 1);
-          }
-          cur += bytecnt;
-        }
-      }
-      if (run & 1) {
-        // Literal batch: must output a multiple of 8, except for the last batch
-        int batch_len_div8;
-        batch_len      = max(min(32, (int)(run >> 1) * 8), 1);
-        batch_len_div8 = (batch_len + 7) >> 3;
-        run -= batch_len_div8 * 2;
-        cur += batch_len_div8 * dict_bits;
-      } else {
-        batch_len = max(min(32, (int)(run >> 1)), 1);
-        run -= batch_len * 2;
-      }
-      s->dict_run   = run;
-      s->data_start = cur;
-      is_literal    = run & 1;
-      __threadfence_block();
-    }
-    __syncwarp();
-    is_literal = shuffle(is_literal);
-    batch_len  = shuffle(batch_len);
-
-    // compute dictionary index.
-    int dict_idx = 0;
-    if (t < batch_len) {
-      dict_idx = s->dict_val;
-      if (is_literal) {
-        int32_t ofs      = (t - ((batch_len + 7) & ~7)) * dict_bits;
-        const uint8_t* p = s->data_start + (ofs >> 3);
-        ofs &= 7;
-        if (p < end) {
-          uint32_t c = 8 - ofs;
-          dict_idx   = (*p++) >> ofs;
-          if (c < dict_bits && p < end) {
-            dict_idx |= (*p++) << c;
-            c += 8;
-            if (c < dict_bits && p < end) {
-              dict_idx |= (*p++) << c;
-              c += 8;
-              if (c < dict_bits && p < end) { dict_idx |= (*p++) << c; }
-            }
-          }
-          dict_idx &= (1 << dict_bits) - 1;
-        }
-      }
-
-      // if we're not computing sizes, store off the dictionary index
-      if constexpr (!sizes_only) { sb->dict_idx[rolling_index(pos + t)] = dict_idx; }
-    }
-
-    // if we're computing sizes, add the length(s)
-    if constexpr (sizes_only) {
-      int const len = [&]() {
-        if (t >= batch_len || (pos + t >= target_pos)) { return 0; }
-        uint32_t const dict_pos = (s->dict_bits > 0) ? dict_idx * sizeof(string_index_pair) : 0;
-        if (dict_pos < (uint32_t)s->dict_size) {
-          const auto* src = reinterpret_cast<const string_index_pair*>(s->dict_base + dict_pos);
-          return src->second;
-        }
-        return 0;
-      }();
-
-      using WarpReduce = cub::WarpReduce<size_type>;
-      __shared__ typename WarpReduce::TempStorage temp_storage;
-      // note: str_len will only be valid on thread 0.
-      str_len += WarpReduce(temp_storage).Sum(len);
-    }
-
-    pos += batch_len;
-  }
-  return {pos, str_len};
-}
-
```

Comment:
This file still contains `gpuDecodeRleBooleans`. So you extract some `gpuDecode*` functions out of here but leave some remaining. Please either extract all.

In addition, you also should consider adding `page_decode.cu` and leave only the declaration of these functions in a `page_decode.hpp` header instead. Including the full function definition in the header will duplicate code and bloat the compiled binary.

Reply:
The moved functions are all device functions.  My (probably out of date) understanding is that absent the use of the `-dc` compiler option, you can't call device functions in a separate compilation unit. I tried doing this initially and got 'Unresolved extern function' errors during compilation. If there is a way to do this please point me to an example and I'll happily convert.

I only moved the functions used by both `page_data.cu` and `page_string_decode.cu` to limit the bloat you reference.

Reply:
> you can't call device functions in a separate compilation unit

Oh sorry I forgot about that. But then we still need to move all them out for consistency. Code bloats only if you call an inline function multiple times. Here we have to do that due to the restriction of device code mentioned above so there should be no difference.